### ISSUE TITLE
Be able to use when expressions in finally Tasks

### ISSUE THREAD

**Issue by bobcatfish** [tektoncd/pipeline#3438]
### Feature request

To be able to combine [when expressions](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#guard-task-execution-using-whenexpressions) with [task execution status](https://github.com/tektoncd/community/pull/234) in finally tasks to be able to express when you want actions to happen in the case of the failure of another Task (without having to modify the Task to be aware of the success/failure case, e.g. use Tasks straight from the catalog).

### Use case

From Andrew Johnson on slack:

> I'd like to conditionally run a set of steps (can be tasks as well) if a task in the main pipeline fails - So an OnFailure (of a specific task) as opposed to Finally.

For example, let's say we want to post to slack (using [this slack catalog task](https://github.com/tektoncd/catalog/tree/master/task/send-to-channel-slack/0.1#post-a-message-to-slack) when a certain Task in the Pipeline fails:

```
  tasks:
    - name: do-things
      taskRef:
        name: do-things
  finally:
    - name: alert-slack
       when:
          - input: "$(context.pipelineRun.Tasks.do-things.status)"
            operator: in
            values: ["failed"]
      taskRef:
        name: send-to-channel-slack
      params:
        - name: channel-id
          value: foo-bar
        ...
```

Andrew also describes how this has to be worked around when we do not have this feature:

> The way I'm currently orchestrating this is by writing a failed status to a workspace file before the task and then removing the failed status after the task. If, when we get to finally, the status is still failed then the steps (or tasks) should run. This currently requires custom tasks that check for the file first.

This means that in order to get this functionality:

1. The executing Task has to be aware of the need to track its status in a workspace
2. The finally Task has to be aware of the meaning of the values written to the workspace

(1) and (2) would prevent a user from being able to use catalog Tasks as is for this use case

**Comment by bobcatfish** [comment]
@jerop @pritidesai  might be relevant to your interests! feel free to close if there is already an issue tracking this

**Comment by jerop** [comment]
> @jerop @pritidesai might be relevant to your interests! feel free to close if there is already an issue tracking this

this is something we've been discussing as next steps after execution status and task results are supported in finally tasks but don't have an issue to track afaik, thanks for opening this...will assign myself 

/assign

**Comment by jerop** [comment]
actively working on this issue, wrote the [WhenExpressions in Finally Tasks TEP](https://github.com/tektoncd/community/blob/master/teps/0045-whenexpressions-in-finally-tasks.md) and opened PR with the implementation of the TEP in https://github.com/tektoncd/pipeline/pull/3738

/remove-lifecycle stale

**Comment by bobcatfish** [comment]
hey @jerop this is done, right?

**Comment by jerop** [comment]
> hey @jerop this is done, right?

yes, it's available in release [v0.21](https://github.com/tektoncd/pipeline/releases/tag/v0.21.0):
- [Design](https://github.com/tektoncd/community/blob/main/teps/0045-whenexpressions-in-finally-tasks.md)
- [Implementation](https://github.com/tektoncd/pipeline/pull/3738)
- [Documentation](https://github.com/tektoncd/pipeline/blob/master/docs/pipelines.md#guard-finally-task-execution-using-whenexpressions)
- [Example](https://github.com/tektoncd/pipeline/blob/master/examples/v1beta1/pipelineruns/pipelinerun-with-when-expressions.yaml)
